### PR TITLE
TypeRegistry constructor must check for duplicate instances

### PR DESCRIPTION
**Pull request by mvorisek**
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | n/a

#### Summary

a) `TypeRegistry` methods like `TypeRegistry::register` checks if the type instance is not already registered.
b) `TypeRegistry::lookupName` method expects the instance to have only one name.

Thus I belive the type registry is expected to not hold duplicate type instances and this PR fixes/asserts that.

Also documented so - https://github.com/doctrine/dbal/blob/3.6.4/src/Types/TypeRegistry.php#L14

### CODE REVIEW

File: tests/Types/TypeRegistryTest.php
Comment (derrabus) [inline tests/Types/TypeRegistryTest.php]: I believe the first constructor call should already be covered by other tests. If that's not the case, please move it into a separate test and run assertions on the constructed instance.

File: src/Types/TypeRegistry.php
Comment (derrabus) [inline src/Types/TypeRegistry.php:27]: While this is technically more correct than the previous behavior, I'm afraid that this change will significantly slow down the constructor because each `register()` call performs a lookup and as you already found out, that lookup is not as fast as it should be.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: tests/Types/TypeRegistryTest.php
Comment (mvorisek) [reply tests/Types/TypeRegistryTest.php]: It is definitely covered, I added `new TypeRegistry(['a' => $newType]);` here to assert that the constructor pass otherwise (with single test) - you want me to remove it?

File: src/Types/TypeRegistry.php
Comment (mvorisek) [reply src/Types/TypeRegistry.php:27]: We will have to mimick the `register()` function otherwise, what do you propose?

Comment (mvorisek) [reply src/Types/TypeRegistry.php:27]: once this bugfix PR is merged, https://github.com/doctrine/dbal/pull/6082 optimization PR is RTM and targetting the same branch, so it will be fast and O(1) (O(N) for whole constructor)

File: tests/Types/TypeRegistryTest.php
Comment (derrabus) [reply tests/Types/TypeRegistryTest.php]: yes please

Comment (mvorisek) [reply tests/Types/TypeRegistryTest.php]: addressed

File: src/Types/TypeRegistry.php
Comment (derrabus) [reply src/Types/TypeRegistry.php:27]: I get that, but it's like you hand me the poison just to sell me the antidote afterwards. I cannot merge this change on its own. However, feel free to move the tests you've developed for this PR over to #6082.

Comment (mvorisek) [reply src/Types/TypeRegistry.php:27]: So you want me to close this PR in favor of #6082 incl. changes in this PR?

I made two PRs for better docs purposes, it should not be a problem, as this PR is functional on it's own, only the later PR will improve the performance.

Comment (derrabus) [reply src/Types/TypeRegistry.php:27]: > So you want me to close this PR in favor of #6082 incl. changes in this PR?

That would be one way to do it, yes.

> I made two PRs for better docs purposes,

Which is a good idea in general.

> it should not be a problem, as this PR is functional on it's own,

No, it will decrease performance significantly which is why I cannot accept it in its current form.

**Comment by mvorisek** [comment]
Landed thru #6082.

File: src/Types/TypeRegistry.php
Comment (mvorisek) [reply src/Types/TypeRegistry.php:27]: > That would be one way to do it, yes.

done, see #6082

Comment (mvorisek) [reply src/Types/TypeRegistry.php:27]: @derrabus may I ask you to review the #6082 PR, please?

**Pull request closed without merging**
